Bayesian Blocks for time series and histograms

jakevdp opened the pull request:
This is an implementation of Bayesian Blocks (as well as a few other histogram bin rules of thumb) for the ``astropy.stats`` package. Code ported and adapted from http://github.com/astroML/astroML/

crawfordsm commented:
Thanks for the contribution @jakevdp!   This would be the right place for it.  The one thing I don't think you need is the underscores before the file names.  They should just be 'bayesian_blocks.py' and 'histogram.py'.  

Once you are ready, I am happy to give it a more thorough review and try it out.

jakevdp commented:
In other packages I work on (e.g. scikit-learn) we often prepend underscores on submodules that are not intended to be used directly by users. In this case, everything users will need from ``_histogram.py`` and ``_bayesian_blocks.py`` is imported into the ``astropy.stats`` namespace; the separate files exist mainly for purposes of code organization rather than for purposes of import hierarchy. Without the underscores, users may get confused about, e.g. the ``astropy.stats.bayesian_blocks`` function vs the ``astropy.stats.bayesian_blocks`` submodule. I find underscores to be helpful for this reason. Is that not a convention used in astropy?

Another question: what level of examples/documentation is expected in astropy for a new feature like this? Are docstrings with examples enough, or should I be planning to add to the narrative documentation in ``astropy/docs``?

One more quick question for @eteq - when folks requested this feature, were they primarily interested in the variable bin width histograms, or were they more interested in the various time series applications of the BB algorithms? Thanks!

jakevdp commented:
Test failures look to be related to doc tests in Python 3. I'll clean those up.

embray commented:
@jakevdp I think that underscore convention is not a bad idea for "internal" organizational modules, but not something we've been doing for Astropy, so until and unless we get fully behind that it would be out of place.  We've mostly enforced what imports to use through the docs.

I'll leave it to the maintainers of the stats package to determine what level of docs is necessary for this, but docstrings with examples is the minimum and often good enough.

crawfordsm commented:
I'd agree with @embray that the minimum requirement would be examples in the doc strings.  As most of the tasks are currently self contained in the stats package, that has been the format that we have been following so far.   

However, if you have more material and want to contribute a page describing the use of the bayesian blocks in the doc page for the stats, I think that would be very helpful to expand on what may be presented in the doc strings.   In the longer term, it is probably what we would want. 

As for what people want, I think the answer is both.  Although my vague recollection was that it was originally brought up in the context of time series, but I maybe wrong.     I'm also cc'ing in @astronomeralex as he was the one taking notes during the session.    If you have examples, maybe this is something or an example that could be given on a doc page?

jakevdp commented:
Thanks. I'll try to put in some more effort later today.

jakevdp commented:
I think things are basically ready for review!

It looks like there are still some issues with the Travis builds. Some of them seem to be due to the SciPy import. Is SciPy not a dependency for astropy?

astrofrog commented:
@jakevdp - thanks for your work on this! Scipy is an optional dependency for Astropy so tests need to be be skipped if it isn't present. Take a look inside the astropy.cosmology tests to see how we usually do that.

jakevdp commented:
Will do – thanks!

jakevdp commented:
Still working on some obscure corner-case errors...

astronomeralex commented:
@jakevdp in response to your question above about what folks are interested in, I agree with @crawfordsm that the answer is both the histograms and time series work. thanks for working on this!

jakevdp commented:
Also, the new code is at ~86% test coverage. Better than the package as a whole, but not as good as it could be :)

I'll work on writing some new tests once I address all the current build failures.

jakevdp commented:
I was thinking about this interface more, and I think I'd like to move more of the logic from the ``bayesian_blocks`` function into the ``FitnessFunc`` base class. Then it would be easier for users to customize how the fitness function is applied... I'll wait until I have tests passing before I do that refactoring.

eteq commented:
@jakevdp - thanks for working on this - great to have this in astropy!

I'm not sure what others think of this (@astrofrog @crawfordsm @astronomeralex) but maybe it also makes sense to add something like the astroml `plotting.hist` function to `astropy.visualization`?  Of course it's not that hard to do it manually with the pieces this PR adds, but it's a nice tool I use sometimes for "quick analysis" histograms.

(Note that if there's agreement to add this, @jakevdp, matplotlib is also an optional dependency like scipy, meaning everything has to be importable even without matplotlib.  The easy solution for that is just to put matplotlib imports *inside* the calling function, rather than at the module level.)

jakevdp commented:
It looks like the test failures now are sphinx errors that I don't understand (probably related to RST formatting issues in doc strings), and the traceback is not all that helpful. Any ideas what might be going wrong?

jakevdp commented:
@eteq – it would be very straightforward to add a histogram plotting utility similar to ``astroML.plotting.hist()``. If I get a couple more +1s, I'll add it :smile:

astronomeralex commented:
@jakevdp +1 to the plotting function

pllim commented:
@jakevdp , regarding Sphinx errors, do you see the same thing when you build the doc locally?

jakevdp commented:
> @jakevdp , regarding Sphinx errors, do you see the same thing when you build the doc locally?

Haven't tried building locally. But the failure only pops up on 1 of the 17 Travis builds.

pllim commented:
That is because only 1 of the 17 builds tests build the doc. My hunch is that you will get a more meaningful error message if you try to build the doc yourself.

astrofrog commented on astropy/stats/bayesian_blocks.py:
```diff
@@ -0,0 +1,424 @@
+        If a string, the following options are supported:
+
+        - 'events' : binned or unbinned event data
+            extra arguments are `p0`, which gives the false alarm probability
```
This should be double back-ticks (will cause a Sphinx failure otherwise)

astrofrog commented on astropy/stats/bayesian_blocks.py:
```diff
@@ -0,0 +1,424 @@
+
+        - 'events' : binned or unbinned event data
+            extra arguments are `p0`, which gives the false alarm probability
+            to compute the prior, or `gamma` which gives the slope of the
```
This should be double back-ticks too - basically single back-ticks are reserved for references that actually resolve to functions or methods.

astrofrog commented on astropy/stats/bayesian_blocks.py:
```diff
@@ -0,0 +1,424 @@
+            to compute the prior, or `gamma` which gives the slope of the
+            prior on the number of bins.
+        - 'regular_events' : non-overlapping events measured at multiples
+            of a fundamental tick rate, `dt`, which must be specified as an
```
As above

astrofrog commented on astropy/stats/bayesian_blocks.py:
```diff
@@ -0,0 +1,424 @@
+        Alternatively, the fitness parameter can be an instance of FitnessFunc
+        or a subclass thereof.
+
+    **kwargs :
```
The ** *may* cause issues, but let's see if you can get things to pass by fixing the single -> double backticks

jakevdp commented:
Ah, got it! Thanks.

astrofrog commented on astropy/stats/bayesian_blocks.py:
```diff
@@ -0,0 +1,424 @@
+
+    References
+    ----------
+    .. [1] Scargle, J `et al.` (2012)
```
the et al doesn't need to be in backticks

astrofrog commented:
By the way, the preferred way to build the docs is to run:

```
python setup.py build_sphinx
```

jakevdp commented:
OK – sphinx build is successful on my machine. I did some of the refactoring I was talking about, and added some more unit tests. Hopefully this last commit will make tests pass!

astrofrog commented:
(just an addendum - the sphinx build has to be successful *and* not emit any warnings in order to pass on Travis)

jakevdp commented:
I'm getting a bunch of sphinx warnings unrelated to this PR:
```
io/votable/references.txt:4: WARNING: py:class reference target not found: SyntaxWarning
io/votable/references.txt:4: WARNING: py:class reference target not found: SyntaxWarning
io/votable/references.txt:4: WARNING: py:class reference target not found: SyntaxWarning
io/votable/references.txt:4: WARNING: py:class reference target not found: RuntimeWarning
io/votable/references.txt:4: WARNING: py:class reference target not found: ValueError
```
Does that mean the build will fail here?

jakevdp commented:
OK – added the ``astropy.visualization.hist()`` routine.

I'm not sure how to unit test this, as it involves a matplotlib API call.

jakevdp commented:
Still seven warnings in sphinx. The first six seem unrelated to this work, and the seventh I can't even begin to decypher:
```
io/votable/references.txt:4: WARNING: py:class reference target not found: SyntaxWarning
io/votable/references.txt:4: WARNING: py:class reference target not found: SyntaxWarning
io/votable/references.txt:4: WARNING: py:class reference target not found: SyntaxWarning
io/votable/references.txt:4: WARNING: py:class reference target not found: RuntimeWarning
io/votable/references.txt:4: WARNING: py:class reference target not found: ValueError
None:1: WARNING: citation not found: R26
```

pllim commented:
For `io/votable/references.txt`, the blame button mentions @mdboom but it had not been changed since 2013. How do all the other PRs pass? Curious.

astrofrog commented:
It could be a Python version thing too - @jakevdp: what Python version are you building the docs with locally?

jakevdp commented:
I'm using 3.4. Just tried in 2.7 and these warnings went away.

Still have the "R26" warning, though... I'll see if I can figure out where that is coming from.

jakevdp commented:
OK – that last commit should make all the tests pass (I hope).

Still TO DO: unit tests for ``astropy.visualization.hist()``. If anyone has ideas on how to best do that, let me know.

astrofrog commented:
@jakevdp - we don't really have 'image' tests for matplotlib plots in the core package yet (though we do that in e.g. WCSAxes). It's tricky because we should be careful about increasing the repository size too much over time with reference images for tests, but I think that will be unavoidable.

In WCSAxes we basically just use the matplotlib test helpers for comparing the figures to reference images.

I'm wondering whether for now you could add a test that just checks that it runs, then open a separate issue for the actual image testing, and we can make decisions there about how to proceed? (just to avoid holding this up).

jakevdp commented:
OK. That makes sense to skip the image things for now. Actually, ``plt.hist()`` has some pretty detailed return values, so I should be able to make some good tests just with those.

Is there anything special I should do to make sure the plot tests don't happen in interactive mode?  Just ``mpl.use('Agg')`` or something like that?

astrofrog commented:
@jakevdp - I *think* the backend should already be taken care of.

jakevdp commented:
OK – I added some non-image-based tests of the histogram code. 

I think this PR is pretty solid now, and I think the travis build should pass once it reaches the top of the queue. If anyone wants to start a review, I'd be grateful – thanks!

crawfordsm commented:
Thanks for getting everything to pass and build.  I'll checkout it out and see if there is anything else that is needed.

jakevdp commented:
One issue I've noticed: if you use IPython and type
```
In [1]: from astropy.visualization import h<TAB>
hist       histogram  
```
the suggestions may be confusing. As a user, which one should be imported to create a histogram?
It's unclear here that ``histogram`` is an internal submodule which the user shouldn't worry about.

In my own packages, I'd fix this by putting an underscore in front of the module name to mark it as internal, but from the above discussion it sounds like devs don't want to use that convention in astropy. An alternative would be to come up with another name that does not convey the notion of histograms, but I'm having trouble coming up with something that doesn't obscure the nature of the code in the submodule. Maybe ``hist_module.py``?

Any suggestions?

jakevdp commented:
Let me know if there's anything else you need from me for this PR.

crawfordsm commented:
My apologies as there was a little work emergency and my time to look at this disappeared.   Taking a look at it now.

jakevdp commented:
No worries. I know from experience that massive PRs like this are pretty hard to review. Feel free to ask if you need clarification on what the algorithm is doing, or choices I've made in the code. Those are probably indications of where I'm missing code comments anyway :smile:

crawfordsm commented on astropy/stats/bayesian_blocks.py:
```diff
@@ -0,0 +1,503 @@
+"""
```
Include astropy license:
# Licensed under a 3-clause BSD style license - see LICENSE.rst

jakevdp replied:
Done

crawfordsm commented on astropy/stats/histogram.py:
```diff
@@ -0,0 +1,332 @@
+"""
```
Include astropy license:

     # Licensed under a 3-clause BSD style license - see LICENSE.rst

jakevdp replied:
Done

crawfordsm commented on astropy/visualization/histogram.py:
```diff
@@ -0,0 +1,60 @@
+        - 'adaptive' : use bayesian blocks for dynamic bin widths
+        - 'blocks' : same as 'adaptive'
+        - 'knuth' : use Knuth's rule to determine bins
+        - 'scott' : use Scott's rule to determine bins
```
Needs to be changed to 'scotts' to make it consistent with what stats.histrogram expects

crawfordsm commented on astropy/visualization/histogram.py:
```diff
@@ -0,0 +1,60 @@
+        array of data to be histogrammed
+
+    bins : int or list or str (optional)
+        If bins is a string, then it must be one of:
```
add an extra line break between each of these so it displays nicely when built by sphinx

jakevdp replied:
Done

crawfordsm commented on astropy/stats/histogram.py:
```diff
@@ -0,0 +1,332 @@
+    a : array_like
+        array of data to be histogrammed
+
+    bins : int or list or str (optional)
```
add an extra line break between each of these so it displays nicely when built by sphinx

jakevdp replied:
You mean between the list items?

crawfordsm replied:
Yes between each of the items so that:

    bins : int or list or str (optional)
        If bins is a string, then it must be one of:

        * 'adaptive' : use bayesian blocks for dynamic bin widths

        * 'blocks' : same as 'adaptive'

        * 'knuth' : use Knuth's rule to determine bins

        * 'scotts' : use Scott's rule to determine bins

        * 'freedman' : use the Freedman-Diaconis rule to determine bins


Otherwise the html sphinx page ends up compress it all down into a single block of text with no itemized list.

jakevdp replied:
Done

crawfordsm commented on astropy/stats/histogram.py:
```diff
@@ -0,0 +1,332 @@
+
+    where :math:`\sigma` is the standard deviation of the data, and
+    :math:`n` is the number of data points.
+
```
Could you add a reference?

crawfordsm commented on astropy/stats/histogram.py:
```diff
@@ -0,0 +1,332 @@
+
+    where :math:`q_{N}` is the :math:`N` percent quartile of the data, and
+    :math:`n` is the number of data points.
+
```
Could you add a reference?

crawfordsm commented on astropy/stats/tests/test_histogram.py:
```diff
@@ -0,0 +1,108 @@
+    range = (0.1, 0.8)
+    
+    for bins in ['scotts', 'freedman', 'blocks', 'adaptive']:
+        print(bins)
```
probably can remove this print statement

crawfordsm commented:
Thanks for understanding. So far, pretty easy to read and using the version in astropy, I've been able to replicate your plots from the astroML site.   Of course, I haven't yet gone into the depths of the bayesian_blocks.py yet,  but I have started with  stats/histogram.py and visualization/histogram.py.   

I would agree with your thought above that perhaps the histogram.py in visualization should be hidden or changed name.  Another option would be to move it to  file named core.py where other similar functions would be included.   However, I think some other people will have opinions about this as well. 

Here are some high level comments that I've noticed so far.  It's late here so I'll continue the review tomorrow, but I wanted to get you these so you'll have a chance to look at them.   I've also add some inline comments as well:

* Is it Scott or Scotts?   You switch between the two between in the docs for hist and histogram.  In the original paper, it  looks like 'Scott'.   I also don't know if you want to make the name consistent with the other functions (ie Knuth vs Knuths), but it might be useful to make these consistent.   At the very least the doc string for hist needs to be changed or made consistent as noted in line. 

* For what I can tell and sorry if I missed something, but it doesn't look like there are any tests for the histogram functions that check that the code returns a range for a known distribution.  It would be useful to add tests for each of the functions just in case something does change, there is at least a test case to compare it to. 

* In the histogram case, there isn't any examples given in the doc string.  These are optional to add, but I think it would be helpful to have them in there or alternatively a page in the documentation describing the functions.   As I said, these are optional, and would be happy to see them added later. 

*  Is there a license on Scargle et al's code?   I see that astroML is BSD which is consistent with astropy but I couldn't easily find a license for Scargle et al.   Just want to make sure that it is compatible (or at the very least we get an okay to include the code in astropy). 

Once again, thanks for contributing the code.  It is a big pull request so it might take some time, but it will be very much worth it having it in astropy!

jakevdp commented:
Thanks for all the comments – spot on, and I'll address all of them. A couple responses:

> Is it Scott or Scotts? 

Probably ``'scott'`` would be most consistent. Good catch!

> Is there a license on Scargle et al's code?

I referred only to Scargle's paper when implementing this: I've never seen Scargle's code beyond the small snippet in the paper appendix, so I think we're OK license-wise.

jakevdp commented:
Thanks @crawfordsm! I think I addressed all the comments so far. I ended up going with ``bins='scott'`` for Scott's rule. I improved tests & documentation as you suggested and also fixed a couple other issues I came across. I added a narrative documentation page with examples of the new histogram plotting functions. I also renamed ``histogram.py`` to ``hist.py`` – that takes care of the IPython tab completion issue well enough.

crawfordsm commented on astropy/stats/histogram.py:
```diff
@@ -0,0 +1,356 @@
+        observed (one-dimensional) data
+    return_bins : bool (optional)
+        if True, then return the bin edges
+
```
Missing description of disp

jakevdp replied:
Updated it – thanks!

crawfordsm commented on astropy/stats/bayesian_blocks.py:
```diff
@@ -0,0 +1,505 @@
+        return t, x, sigma
+
+    def p0_prior(self, N, Ntot):
+        # eq. 21 from Scargle 2012
```
Okay, this is different from what appears in the arXiv version and in the ApJ version (which are different from each other).   Can you add a comment here about that?

jakevdp replied:
This was addressed in this astroML pull request: https://github.com/astroML/astroML/pull/20
The ApJ paper has an error.

jakevdp replied:
Added a commit with this info in the doc string.

crawfordsm commented on astropy/stats/bayesian_blocks.py:
```diff
@@ -0,0 +1,505 @@
+    Examples
+    --------
+    Event data:
+
```
Can you add what the expected output would be from these examples below each one?

jakevdp replied:
I left that out because I thought it would be distracting. The numbers won't mean all that much to the reader of the documentation, and the separate unit tests take care of input/output testing.

jakevdp replied:
I added a unit test with the output from each of these.

crawfordsm commented:
So here are a few more things that I've picked up as I've gone through the code and compared it to previous results.   I can produce separate checks for Scott and Freedman, but I am having a little more trouble with Knuth and the Bayesian Blocks. 

The implementation and optimization of Knuth is slightly different than that in the paper as far as I can tell.   I tried to reproduce one of the results in the paper (using the Abolone data).  For me, it finds a local minimum near the first guess based on the Freedman bins (around 37) and not  the true minimum (around 16).   There is a plateau in the distribution of M values, so this isn't very surprising, but it does behave differently than that in the paper.    I am not sure if this should be changed or just noted in the documentation.   

I'm still working on a comparison for the Bayesian blocks with the original Scargle work, but I'd like to make sure that I am doing the right thing before commenting further.  Have you run your code on any similar data sets as other code?    In particular, I was trying to reproduce their Fig 7 from Scargle et al.  I can reproduce the regular histogram but I seem to get some odd binnings if I use bayesian_blocks.

However, the one thing I noticed was there wasn't much testing of the individual fitness functions in the test suite for bayesian blocks.   It does appear that all the testing is done at the higher level of bayesian_blocks.  If the FitnessFuncs are meant to be used on their own, it might be good to separate out the testing on the lower level of individual functions rather than having all tests run on bayesian_blocks only.  

Anyway thanks for the quick updates and apologies for the slow review.

jakevdp commented:
Thanks for the thorough tests!

The fitness functions are not really meant to be used on their own. Originally most of the stuff from the ``FitnessFunc`` base class was in the ``bayesian_blocks`` function, but I moved it so that the code would be more easily extensible.

Regarding the Knuth result: it's probably an issue of numerical optimization. If we use a different starting guess, we may end up at a local minimum. I'm not sure how to address that, as it's pretty much an unavoidable feature of any black box optimization routine.

jakevdp commented:
Note that the most recent Travis failure is unrelated to this PR: it's a timeout when fetching data used to build the documentation examples.

embray commented:
Hmm, not to get too far off topic, but I didn't realize we had an example in the docs that relies on online resources.  I wonder if we could make a version of that that uses some cached data included in the docs, unless that would distract from the intent of the example too much.

mwcraig commented:
The Travis fail is unrelated to this PR -- an example from SkyCoords timed out on a position lookup `from_name` which generated a sphinx warning.

astrofrog commented:
@mwcraig - restarted

eteq commented:
@crawfordsm - are you still testing/reviewing (wasn't sure from your post above), or are you good to go on this?

cdeil commented on docs/visualization/histogram.rst:
```diff
@@ -0,0 +1,118 @@
+Bayesian Models
+---------------
+
+Though rules-of-thumb like Scott's rule and the Freedman-Diaconis rule are fast and convenient, their strong assumptions about the data make them suboptimal for more complicated distributions. Other methods of bin selection use fitness functions computed on the actual data to choose an optimal binning. Astropy implements two of these examples: Knuth's rule (implemented in :func:`~astropy.stats.kn ...
```
Could you please hard-wrap all text to ~ 80 characters?
(If I'm not mistaken we do this everywhere in Astropy and it helps with git diffs)

cdeil commented:
Unrelated test failure from `astropy.vo.samp`:
https://travis-ci.org/astropy/astropy/jobs/64031597#L2187

Can someone please restart that build?
And please skip that test as long as it's flaky?

Otherwise :+1: , this looks great!

astrofrog commented:
@cdeil - the SAMP failure is different to ones I've seen before, and is actually related to the ``open-files`` flag we are passing - for some reason it thinks that ``/etc/hosts`` has not been closed. I've see that failure in other tests before. @mdboom @embray, I think one of you wrote the ``--open-files`` plugin - can we add a hard-coded exception for ``/etc/hosts`` since that may be beyond our control?

embray commented:
@astrofrog Can you open a separate issue to discuss that so we don't clutter Jake's PR with any more OT discussion than there already is? :)

crawfordsm commented:
Sorry to get back to this so late, but yes, this passes all my checks.

I have two final comments for @jakevdp:
* Can you add a link in docs/stats/index.rst to your new histogram page?
* Can you add an entry into the change log?

After the change log entry, I'd be happy to see this merged in.

jakevdp commented:
Tests weren't running after f2b335205d9bdf0406f9630da55ab7c0b13e2ac3 because of a merge conflict with recent changes in master; I pulled in current master to make tests run.

jakevdp commented:
Not sure how these things are handled in astropy (keeping the branches & merges? rebasing? squashing?) but I'm happy to help however you'd like.

crawfordsm commented:
@embray do you have a specific suggests for how to handle the merge in the commits?

eteq commented:
@jakevdp @crawfordsm - our usual stance is to always rebase against master if there's a merge conflict. If that's particularly onerous for some reason sometimes a merge is ok, but rebase is the standard mostly because it makes for easier-to-understand commit trees.

eteq commented:
Actually, it looks like post-merge a ``git rebase master`` did the job locally, so hopefully that's all you'll have to do @jakevdp - then if the tests pass we can hit the button!

jakevdp commented:
If I rebase, I won't be able to push to this branch. Would you like me to open a new PR with the rebased commits?

mwcraig commented:
@jakevdp -- it is fine to do a force push to your branch.

cdeil commented on astropy/stats/histogram.py:
```diff
@@ -0,0 +1,358 @@
+
+    knuthF = _KnuthF(data)
+    dx0, bins0 = freedman_bin_width(data, True)
+    M0 = len(bins0) - 1
```
Unused variable `M0` - Remove this line?

jakevdp replied:
Done in  f9cdea5 – thanks!

jakevdp commented:
OK – I rebased on master and force-pushed.

crawfordsm commented:
Okay, unless any one has further comments, I'll go ahead and merge this in end of today.  Thank you very much @jakevdp for the contribution.

eteq commented:
My only contribution to add is that "force push" always makes me think of Star Wars.

(Also, thanks for all of your work on this @jakevdp !)

jakevdp commented:
Thanks everyone!

cdeil commented:
I'll use the Bayesian block histogram binning for my work now.
This is fantastic, thanks!

cdeil commented:
@astrofrog or @eteq – This is not showing up in the docs on readthedocs yet (there should be a page e.g. here: http://astropy.readthedocs.org/en/latest/visualization/histogram.html). It's a bit weird because the last html build passed: https://readthedocs.org/builds/astropy/2888032/. Probably you have to wipe and start a clean build?

cdeil commented:
I'm `deil` on RTD.

astrofrog commented:
@cdeil - you are now an admin for Astropy on RTD :) I tried to wipe and restart build so let's see what happens.

cdeil commented:
The build on RDT passed: https://readthedocs.org/builds/astropy/2891064/

But the docs pages still don't show the extra page or functions defined here!?

Sometimes pages update on RTD hours later. I think their status info doesn't always tell what's going on. Probably best to just wait until tomorrow and see if it's fixed by then.

cdeil commented:
I wiped and restarted the RTD build again.
This time it worked, i.e. the docs for this are now online:
http://astropy.readthedocs.org/en/latest/visualization/histogram.html

astrofrog commented:
@cdeil - thanks!
